File: HWs/HW5_IncompleteLCG/Python/main.py
from math import sqrt
from math import inf
from functools import reduce
from operator import mul


def prime_factors(n):
    factors = []
    while n % 2 == 0:
        factors.append(2)
        n //= 2
    for i in range(3, int(sqrt(n)+1), 2):
        while n % i == 0:
            factors.append(i)
            n //= i

    if n > 2:
        factors.append(n)
    return factors


def gcd(a, b):
    a, b = (a, b) if a >= b else (b, a)
    while a % b > 0:
        a, b = b, a % b
    return b


def pos_mod(a, m):
    tmp = a % m
    return tmp if tmp >= 0 else tmp + m


def inverse(a, n):
    t, r, newt, newr = 0, n, 1, a
    while newr != 0:
        q = r // newr
        t, newt = newt, t - (q * newt)
        r, newr = newr, r - (q * newr)
    if r > 1:
        raise ValueError("not invertible")
    if t < 0:
        t += n
    return t


def find_x1(A, M, inv, x2, x3):
    C = (x3 - A*x2) % M
    shift = (x2 - C) % M
    x1 = pos_mod((inv * shift), M)
    return x1


if __name__ == '__main__':
    [M, x2, x3] = [int(i) for i in input().split()]
    factors = prime_factors(M)
    candidate = reduce(mul, set(factors))
    cnt = 0
    x1_max = -inf
    x1_min = inf
    if M % 4 == 0:
        candidate *= gcd(candidate, 4)
    for A in range(candidate+1, M, candidate):
        inv = inverse(A, M)
        x1 = find_x1(A, M, inv, x2, x3)
        if x1_max < x1:
            x1_max = x1
        if x1_min > x1:
            x1_min = x1
        cnt += 1
    print(f'{cnt} {x1_min} {x1_max}')

File: HWs/HW5_IncompleteLCG/cpp/main.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <utility>
#include <tuple>

long long int inverse(long long int a, long long int m);

long long int find_x1(long long int A, long long int M, long long int x2, long long int x3);

using namespace std;

/**
 * Return prime factors, each prime included only once
 * @param n - number to factorize
 * @return vector of prime factors
 */
vector<long long int> prime_factors_set(long long int n) {
    vector<long long int> factors;
    auto handlePossibleFactor = [&](long long int f) {
        if (n % f == 0) {
            factors.push_back(f);
            n /= f;
        }
        while (n % f == 0) {
            n /= f;
        }
    };

    handlePossibleFactor(2);
    long long int stop = (long long int) sqrt(n)+1;
    for (long long int i = 3;  i < stop; i+= 2){
        handlePossibleFactor(i);
    }
    if (n > 2) {
        factors.push_back(n);
    }
    return factors;
}

long long int gcd(long long int a, long long int b) {
    if (a < b) {
        swap(a, b);
    }
    while (a % b > 0) {
        tie(a, b) = make_pair(b, a % b);
    }
    return b;
}

int main() {
    ios_base::sync_with_stdio(false);
    long long int M, x2, x3;
    cin >> M >> x2 >> x3;
    auto factors = prime_factors_set(M);
    long long int candidate = 1;
    for (auto p : factors) {
        candidate *= p;
    }
    if (M % 4 == 0) {
//        candidate *= gcd(candidate, 4);
        candidate *= 2; //if M divisible by 4, candidate already constructed by multiplying 2, so just multiply again
    }

    long long int x1_max = -1;
    int64_t x1_min = INT64_MAX;
    uint64_t count = 0;

    for (long long int A = candidate+1; A < M; A+=candidate) {
        long long int x1 = find_x1(A, M, x2, x3);
        if (x1_max < x1) {
            x1_max = x1;
        }
        if (x1_min > x1) {
            x1_min = x1;
        }
        count++;
    }

    cout << count << " " << x1_min << " " << x1_max << endl;

    return 0;
}

long long int find_x1(long long int A, long long int M, long long int x2, long long int x3) {
    long long int C = (x3 - A*x2) % M;
    long long int shift = (x2 - C) % M;
    long long int x1 = (inverse(A, M) * shift) % M;

    return x1 < 0 ? x1 + M : x1;
}

long long int inverse(long long int a, long long int m) {
    long long int t = 0, r = m, newt = 1, newr = a;
    while (newr != 0) {
        long long int q = r / newr;
        tie(t, newt) = make_pair(newt, t - (q * newt));
        tie(r, newr) = make_pair(newr, r - (q * newr));
    }
    if (t < 0) {
        t += m;
    }
    return t;
}
